services/src/users/postgres_userdb.rs

```diff
@@ -0,0 +1,175 @@
+use crate::error;
+use crate::error::Result;
+use crate::users::session::{Session, SessionToken};
+use crate::users::user::User;
+use crate::users::user::{UserCredentials, UserId, UserRegistration};
+use crate::users::userdb::UserDB;
+use crate::util::identifiers::Identifier;
+use crate::util::user_input::Validated;
+use async_trait::async_trait;
+use bb8_postgres::bb8::Pool;
+use bb8_postgres::tokio_postgres::error::SqlState;
+use bb8_postgres::tokio_postgres::NoTls;
+use bb8_postgres::PostgresConnectionManager;
+use snafu::ResultExt;
+
+struct PostgresUserDB {
+    conn_pool: Pool<PostgresConnectionManager<NoTls>>, // TODO: support Tls connection as well
+}
+
+impl PostgresUserDB {
+    async fn new(conn_pool: Pool<PostgresConnectionManager<NoTls>>) -> Result<Self> {
+        let a = Self { conn_pool };
+        a.update_schema().await?;
+        Ok(a)
+    }
+
+    async fn schema_version(&self) -> Result<i32> {
+        // TODO: handle if version table not exists
+        let conn = self.conn_pool.get().await.map_err(|_| error::Error::BB8)?; // TODO: error reporting
+
+        let stmt = match conn.prepare("SELECT version from version").await {
+            Ok(stmt) => stmt,
+            Err(e) => {
+                if let Some(code) = e.code() {
+                    if *code == SqlState::UNDEFINED_TABLE {
+                        // TODO: log
+                        println!("UserDB: Uninitialized schema");
+                        return Ok(0);
+                    }
+                }
+                return Err(error::Error::TokioPostgres { source: e });
+            }
+        };
+
+        let row = conn
+            .query_one(&stmt, &[])
+            .await
+            .context(error::TokioPostgres)?;
+
+        Ok(row.get::<usize, i32>(0))
+    }
+
+    async fn update_schema(&self) -> Result<()> {
+        let mut version = self.schema_version().await?;
+
+        let conn = self.conn_pool.get().await.map_err(|_| error::Error::BB8)?; // TODO: error reporting
+        loop {
+            match version {
+                0 => {
+                    conn.batch_execute(
+                        "\
+                        CREATE TABLE version (\
+                            version INT\
+                        );\
+                        INSERT INTO version VALUES (1);\
+
+                        CREATE TABLE users (
+                            id UUID PRIMARY KEY,
+                            email character varying (256) UNIQUE,
+                            password_hash character varying (256),
+                            real_name character varying (256)
+                        );
+                        ",
```

Comment:
Looks like a reasonable approach, maybe can put the instructions into *.sql files and include them here with `include_str!`.

Reply:
what's the advantage?

Reply:
Syntax-highlighting and the chance to execute them directly more easily.
Otherwise, I would at least use raw strings instead of escaping all lines, etc,

Reply:
Future schema changes might be only single lines. creating separate files doesn't scale well. The new line escapes are no longer there.

Reply:
I hope that schema changes are not a scalability issue :o. 

You could take a final look at diesel https://diesel.rs/guides/getting-started/, which has migration scripts in files, and then we conclude this comment.

Reply:
I meant that its more easy to keep lots of small schema changes in one file than having them all in individual files. We can change this easily later on anyway. For now I vote for keeping the statments in the source